ocaml/xapi/xapi_observer_components.ml

```diff
@@ -54,8 +56,22 @@ let assert_valid_components components =
   with Unsupported_Component component ->
     raise Api_errors.(Server_error (invalid_value, ["component"; component]))
 
+let filter_out_exp_components components =
+  let open Xapi_globs in
+  let component_set = components |> List.map to_string |> StringSet.of_list in
```

Comment:
We could have a set for type t instead, although I think you had problems with circular dependencies that way?

Reply:
Yes, using the custom setter to make a set of type t creates a circular dependency. Although after adding the strings to `constants.ml` we can validate the elements of the string set.